### PR TITLE
Rchain-3819 Broadcast block meta-data instead of the whole block

### PR DESCRIPTION
**Pull request by marcin-rzeznicki**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

Closes https://rchain.atlassian.net/browse/RCHAIN-3819
Block are no longer streamed to peers, instead the block hash is sent. If peers decide to process this block, they request it by hash.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

https://rchain.atlassian.net/browse/RCHAIN-3819

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

- `Cell` typeclass needed to be overhauled. There were numerous problems with it:

1. `MVar` based cell was prone to deadlock on error or cancelation (fixed),
1. `Ref` cell did not behave the same way as `MVar` cell. It still doesn't because it is impossible to implement atomic `S => F[S]` using Cell only, but since most users did not have any problems with it then it means that they simply do **not** need guarantees that `Cell` has (or, was supposed to have, see(1)). `Ref` cell needs to be eventually deleted. All its users can safely migrate to cats-effect `Ref` (ping @lukasz-golebiewski )
1. Production code used the so-called `unsafe` cell - which was, to put it mildly, **very** unsafe. Deleted and reimplemented on top of `Ref` (see 2)

- Tests in casper are extremely brittle. They will be subject to further re-writes to something better. Currently new methods were introduced (`propagate`, `syncWith`) which may ease the pain, nonetheless they're not a solution to the problem

- Reviewers are asked to pay extreme attention when reviewing tests. Two hopeless cases are now `ignored`:
1. _reject addBlock when there exist deploy by the same (user, millisecond timestamp) in the chain_ ticket that was supposed to fix it is closed, yet this test does not pass, investigate further
2. _handle a long chain of block requests appropriately_ it doesn't make much sense in the presence of hashes and why on earth the only thing it tests is logs? (ping @dzajkowski )

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by marcin-rzeznicki** [comment]
bors try

File: models/src/main/scala/coop/rchain/casper/protocol/PacketTypeTag.scala [hunk 1]
```diff
@@ -58,14 +58,37 @@ object PacketTypeTag extends Enum[PacketTypeTag] {
 
 import coop.rchain.casper.protocol.PacketTypeTag.ValueOf
 
+sealed abstract class PacketParseResult[+A](val isSuccess: Boolean) {
+  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
+  final def get: A   = fold(identity)(err => throw new NoSuchElementException(err))
+  final def toEither = fold(Right(_): Either[String, A])(Left(_))
```
Comment (sebtomba) [inline models/src/main/scala/coop/rchain/casper/protocol/PacketTypeTag.scala:64]: Maybe we can use here cats `.asRight` and `.asLeft` for a more readable code?

**Review by sebtomba — APPROVED** [review]

File: casper/src/test/scala/coop/rchain/casper/MultiParenCasperFinalizationSpec.scala [hunk 2]
```diff
@@ -1,58 +0,0 @@
...
-
-import cats.implicits._
-import coop.rchain.casper.helper.TestNode
-import coop.rchain.casper.helper.TestNode._
-import coop.rchain.shared.scalatestcontrib._
-import coop.rchain.casper.util.ConstructDeploy
-import coop.rchain.p2p.EffectsTestInstances.LogicalTime
-import monix.execution.Scheduler.Implicits.global
-import org.scalatest.{FlatSpec, Inspectors, Matchers}
-
```
Comment (EncodePanda) [inline casper/src/test/scala/coop/rchain/casper/MultiParenCasperFinalizationSpec.scala:11]: Lol

**Review by EncodePanda — APPROVED** [review]
Good catch on the Cell stuff. I'm happy we have you here 👍❤️

File: casper/src/test/scala/coop/rchain/casper/MultiParenCasperFinalizationSpec.scala (hunk 2 above)
Comment (marcin-rzeznicki) [reply casper/src/test/scala/coop/rchain/casper/MultiParenCasperFinalizationSpec.scala:11]: Spring cleaning :smile:

**Comment by marcin-rzeznicki** [comment]
Thank you @rabbitonweb for your kind words!

**Pull request merged**